Add `string` RHS function

garfieldnate opened the pull request:
Mostly this a copy of the trim function with most of the code deleted; we don't need to do any manual copying or freeing of strings, since `make_str_constant` does this for us.

Add a new category of unit tests, `BuiltinRHSTests`; I do see some existing tests that exercise the built-in RHS functions, but they are in the `ChunkingTests` category, and we're not testing chunks here.

Fixes #313.

garfieldnate committed:
Add `string` RHS function

Mostly this a copy of the trim function with most of the code deleted;
we don't need to do any manual copying or freeing of strings, since
`make_str_constant` does this for us.

Add a new category of unit tests, `BuiltinRHSTests`; I do see some
existing tests that exercise the built-in RHS functions, but they are
in the `ChunkingTests` category, and we're not testing chunks here.

Fixes #313.

marinier commented:
This appears to work, but one possible issue is that it doesn't capture many digits. It seems to have the same problem as #331. For example, consider this rule I used for testing:

```
sp {test
(state <s> ^superstate nil)
-->
(<s> ^a (string 123456789) ^b (string (/ 1 9)) ^c (string -123456789) ^d (string -1.123456789) ^e (float (string -1.123456789)))}
```

The `b` string doesn't print as many digits as it could, and the `d` string is actually truncated. Furthermore, the `e` string shows that converting to a string and back to a float does not give you the original float.

I think it just needs the same fix as #331.

garfieldnate committed:
Preserve full decimal precision in string RHS

We add a new optional parameter to Symbol::to_string which specifies
the number of decimal places to use when converting a float to a string.
By default, this is 6, which is the same as used by fprintf, etc.
We only cache stringified floats when the default precision is used,
since that is the common case throughout the codebase.

Update the tests to check that a round-trip from float to string and
back to float again preserves the original value.

garfieldnate committed:
Use std::fixed to preserve precision

We want to preserve the old symbol float printing behavior, which always
had 6 digits after the decimal. To do this, we must set the formatting
to fixed. This also ensures that we are specifying the number of
digits *after* the decimal, not just the total digits.

Apply the same `std::fixed` fix to `AddDoubleInputWME`.

Add several more decimal points to the tests for the `string` RHS, now
that we can correctly round-trip them. 18 digits is more like it!
